docarray/array/storage/elastic/backend.py

```diff
@@ -0,0 +1,217 @@
+import copy
+import itertools
+import warnings
+from dataclasses import dataclass, field
+from typing import (
+    Dict,
+    Optional,
+    TYPE_CHECKING,
+    Union,
+    Tuple,
+    List,
+    Sequence,
+    Generator,
+    Iterator,
+)
+
+import numpy as np
+import uuid
+
+from ..base.backend import BaseBackendMixin
+from .... import DocumentArray, Document
+from ....helper import dataclass_from_dict
+
+from elasticsearch import Elasticsearch
+from elasticsearch.helpers import bulk
+
+if TYPE_CHECKING:
+    from ....types import (
+        DocumentArraySourceType,
+    )
+    from ....types import DocumentArraySourceType, ArrayType
+
+from docarray.math.helper import EPSILON
+
+
+def _sanitize_table_name(table_name: str) -> str:
+    ret = ''.join(c for c in table_name if c.isalnum() or c == '_')
+    if ret != table_name:
+        warnings.warn(f'The table name is changed to {ret} due to illegal characters')
+    return ret
+
+
+@dataclass
+class ElasticConfig:
+    n_dim: int  # dims  in elastic
+    basic_auth: Optional[Tuple[str, str]] = None
+    ca_certs: Optional[str] = None
+    distance: str = 'l2_norm'  # similarity in elastic
+    host: Optional[str] = field(default='http://localhost')
+    port: Optional[int] = field(default=9200)
+    index_name: Optional[str] = field(default=None)
+    serialize_config: Dict = field(default_factory=dict)
+
+
+class BackendMixin(BaseBackendMixin):
+    """Provide necessary functions to enable this storage backend."""
+
+    def _init_storage(
+        self,
+        _docs: Optional['DocumentArraySourceType'] = None,
+        config: Optional[Union[ElasticConfig, Dict]] = None,
+        **kwargs,
+    ):
+
+        config = copy.deepcopy(config)
+        if not config:
+            raise ValueError('Empty config is not allowed for Elastic storage')
+        elif isinstance(config, dict):
+            config = dataclass_from_dict(ElasticConfig, config)
+
+        if config.index_name is None:
+            self._persist = False
+            id = uuid.uuid4().hex
+            config.index_name = 'index_name__' + id
+        else:
+            self._persist = True
+
+        self._index_name_offset2id = 'offset2id__' + config.index_name
+        self._config = config
+        self.n_dim = self._config.n_dim
+        self._client = self._build_client()
+        self._build_offset2id_index()
+
+        # Note super()._init_storage() calls _load_offset2ids which calls _get_offset2ids_meta
+        super()._init_storage()
+
+        if _docs is None:
+            return
+        elif isinstance(
+            _docs, (DocumentArray, Sequence, Generator, Iterator, itertools.chain)
```

Comment:
`if isinstance(docs, Iterable) and self._ref_doc is not None:` is good enough here.

Reply:
This would allow other iterables that are not supported? I leave it as is, which is how it is done in other storage backends.

Reply:
Please check the latest main, this is the current implementation in other storage